src/dsp_tools/utils/xmlupload/resource_create_client.py

```diff
@@ -0,0 +1,379 @@
+        if not start_era:
+            start_era = "CE"
+        if end_date and not end_era:
+            end_era = "CE"
+
+    year, month, day = _parse_date(start_date)
+    res["knora-api:dateValueHasStartYear"] = int(year)
+    if month:
+        res["knora-api:dateValueHasStartMonth"] = int(month)
+    if day:
+        res["knora-api:dateValueHasStartDay"] = int(day)
+    if end_era:
+        res["knora-api:dateValueHasEndEra"] = end_era
+
+    if calendar:
+        res["knora-api:dateValueHasCalendar"] = calendar
+    if start_era:
+        res["knora-api:dateValueHasStartEra"] = start_era
+    if end_date:
+        year, month, day = _parse_date(end_date)
+        res["knora-api:dateValueHasEndYear"] = int(year)
+        if month:
+            res["knora-api:dateValueHasEndMonth"] = int(month)
+        if day:
+            res["knora-api:dateValueHasEndDay"] = int(day)
+    return res
+
+
+def _parse_date(date: str) -> tuple[str, str | None, str | None]:
+    date_pattern = r"(\d{4})(?:-(\d{1,2}))?(?:-(\d{1,2}))?"
+    date_match = regex.search(date_pattern, date)
+    assert date_match
+    match date_match.groups():
+        case (year, month, day):
+            assert year
+            return year, month, day
+    raise UserError(f"Could not parse date: {date}")
+
+
+def _make_decimal_value(value: XMLValue) -> dict[str, Any]:
+    assert isinstance(value.value, str)
+    return {
+        "@type": "knora-api:DecimalValue",
+        "knora-api:decimalValueAsDecimal": {
+            "@type": "xsd:decimal",
+            "@value": str(float(value.value)),
+        },
+    }
+
+
+def _make_geometry_value(value: XMLValue) -> dict[str, Any]:
+    assert isinstance(value.value, str)
+    encoded_value = json.dumps(json.loads(value.value))
+    return {
+        "@type": "knora-api:GeomValue",
+        "knora-api:geometryValueAsGeometry": encoded_value,
+    }
+
+
+def _make_geoname_value(value: XMLValue) -> dict[str, Any]:
+    return {
+        "@type": "knora-api:GeonameValue",
+        "knora-api:geonameValueAsGeonameCode": value.value,
+    }
+
+
+def _make_integer_value(value: XMLValue) -> dict[str, Any]:
+    assert isinstance(value.value, str)
+    return {
+        "@type": "knora-api:IntValue",
+        "knora-api:intValueAsInt": int(value.value),
+    }
+
+
+def _make_interval_value(value: XMLValue) -> dict[str, Any]:
+    match value.value:
+        case str() as s:
+            start, end = tuple(s.split(":", 1))
+            return {
+                "@type": "knora-api:IntervalValue",
+                "knora-api:intervalValueHasStart": {
+                    "@type": "xsd:decimal",
+                    "@value": str(float(start)),
+                },
+                "knora-api:intervalValueHasEnd": {
+                    "@type": "xsd:decimal",
+                    "@value": str(float(end)),
+                },
+            }
+        case _:
+            raise UserError(f"Unexpected interval value: {value.value}")
```

Comment:
This one just slipped my attention, according to my own system, it should just be an assert.  
I was under the impression that `UserError` is thrown if something is "the user's fault",e.g. invalid input; whereas `BaseError` was something beyond the control of the user (so, similar to the distinction between 400 and 500 error range in HTTP). Is that not how the error system works in DSP-TOOLS?

Reply:
I think these 2 ways of thinking about UserError are both valid. But from a pragmatic point of view, we should make sure that UserErrors have enough context info in their message, so that an RDU user can take action. It would also be possible to raise a BaseError here, catch it further up in the hierarchy, and turning it into a nice UserError.

Anyways, `"Unexpected interval value: xy"` doesn't seem to be user-friendly enough, plus it misses important context information (which resource/property is affected?)